File: starngepainter/code.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <map>
#include <unordered_set>
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    bool isPrintable(vector<vector<int>> &grid)
    {
        // Step 1: Finding Distant Colors in the Grid || ex: 1, 1, 1, 1 -> 1
        unordered_set<int> st; // Set to store distinct colors in the grid
        int n = grid.size(), m = grid[0].size();

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                st.insert(grid[i][j]); // Insert all colors in the grid into the set 'st' || ex: 1, 2, 2, 1 -> 1, 2
            }
        }

        // Step 2: Finding Coordinates of Colors
        map<int, vector<int>> row, col; // Maps to store the starting and ending row and column coordinates for each color

        for (auto it : st)
        {
            int mn1 = n, mn2 = m, mx1 = 0, mx2 = 0; // Variables to store the starting and ending row and column coordinates for each color

            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < m; j++)
                {
                    if (grid[i][j] == it) // If the color at (i, j) is equal to 'it', update the coordinates || ex: 1, 1, 1, 1 -> 1, 4, 1, 4
                    {
                        mn1 = min(i, mn1); // Update the coordinates || ex: 1, 1, 1, 1 -> 1, 4, 1, 4
                        mx1 = max(i, mx1);
                        mn2 = min(j, mn2);
                        mx2 = max(j, mx2);
                    }
                }
            }
            row[it].push_back(mn1); // Add the coordinates to the maps
            row[it].push_back(mx1);
            col[it].push_back(mn2);
            col[it].push_back(mx2);
        }

        // Step 3: Building the Dependency Graph
        vector<vector<int>> graph(61, vector<int>()); // 2D vector representing the dependency graph between colors
        int indeg[61];                                // Array to store the in-degree of each color
        memset(indeg, 0, sizeof indeg);

        for (auto it : st)
        {
            int mnr = row[it][0], mxr = row[it][1], mnc = col[it][0], mxc = col[it][1]; // Variables to store the starting and ending row and column coordinates for color 'it' || ex: 1, 1, 1, 1 -> 1, 4, 1, 4
            set<int> st2;                                                               // Set to store colors that are inside the rectangle of color 'it'

            for (int i = mnr; i <= mxr; i++)
            {
                for (int j = mnc; j <= mxc; j++)
                {
                    if (it != grid[i][j])
                        st2.insert(grid[i][j]); // Insert all colors inside the rectangle of color 'it' into the set 'st2'
                }
            }
            indeg[it] += st2.size(); // Increment the in-degree of color 'it' based on the number of colors inside its rectangle
            for (auto pt : st2)
                graph[pt].push_back(it); // Add edges from colors inside the rectangle to color 'it'
        }

        // Step 4: Standard Topological Sort
        int cnt = 0;  // Variable to count the number of colors processed during topological sort
        queue<int> q; // Queue for the BFS-based topological sort

        for (int i = 1; i <= 60; i++)
        {
            if (st.find(i) != st.end() && indeg[i] == 0)
                q.push(i); // Add colors with in-degree 0 to the queue to start the topological sort
        }

        while (!q.empty())
        {
            int x = q.front();
            q.pop();
            cnt++;

            for (auto it : graph[x])
            {
                indeg[it]--;
                if (indeg[it] == 0)
                    q.push(it); // If the in-degree of a color becomes 0, add it to the queue for processing
            }
        }

        // Step 5: Checking the Result
        // If the 'cnt' variable is equal to the number of distinct colors in the 'targetGrid',
        // it means all colors were successfully processed in the topological sort, and we can return true.
        // Otherwise, we return false.
        return cnt == st.size();
    }
};

int main()
{
    Solution s;
    vector<vector<int>> grid = {{1, 1, 1, 1},
                                {1, 2, 2, 1},
                                {1, 2, 2, 1},
                                {1, 1, 1, 1}};
    cout << s.isPrintable(grid) << endl;
    return 0;
}

File: starngepainter/code.rs
use std::collections::{HashSet, HashMap, VecDeque};

impl Solution {
    pub fn is_printable(grid: Vec<Vec<i32>>) -> bool {
        // Step 1: Finding Distant Colors in the Grid
        // Create a HashSet 'st' to store distinct colors in the grid.
        let mut st: HashSet<i32> = HashSet::new();
        let n = grid.len();
        let m = grid[0].len();
        for i in 0..n {
            for j in 0..m {
                st.insert(grid[i][j]);
            }
        }

        // Step 2: Finding Coordinates of Colors
        // Create HashMaps 'row' and 'col' to store the starting and ending row and column coordinates for each color.
        let mut row: HashMap<i32, Vec<i32>> = HashMap::new();
        let mut col: HashMap<i32, Vec<i32>> = HashMap::new();

        for it in &st {
            let mut mn1 = n as i32;
            let mut mx1 = 0;
            let mut mn2 = m as i32;
            let mut mx2 = 0;

            // Iterate through the grid to find the starting and ending row and column coordinates for color 'it'.
            for i in 0..n {
                for j in 0..m {
                    if grid[i][j] == *it {
                        mn1 = mn1.min(i as i32);
                        mx1 = mx1.max(i as i32);
                        mn2 = mn2.min(j as i32);
                        mx2 = mx2.max(j as i32);
                    }
                }
            }

            // Store the coordinates in the HashMaps.
            row.entry(*it).or_insert(Vec::new()).push(mn1);
            row.get_mut(it).unwrap().push(mx1);
            col.entry(*it).or_insert(Vec::new()).push(mn2);
            col.get_mut(it).unwrap().push(mx2);
        }

        // Step 3: Building the Dependency Graph
        // Create a 2D vector 'graph' to represent the dependency graph between colors.
        let mut graph: Vec<Vec<i32>> = vec![vec![]; 61];
        let mut indegree: [i32; 61] = [0; 61];

        for it in &st {
            let mnr = row[it][0];
            let mxr = row[it][1];
            let mnc = col[it][0];
            let mxc = col[it][1];

            // Create a HashSet 'st2' to store colors that are inside the rectangle of color 'it'.
            let mut st2: HashSet<i32> = HashSet::new();
            for i in mnr..=mxr {
                for j in mnc..=mxc {
                    if *it != grid[i as usize][j as usize] {
                        st2.insert(grid[i as usize][j as usize]);
                    }
                }
            }

            // Increment the in-degree of color 'it' based on the number of colors inside its rectangle.
            indegree[*it as usize] = st2.len() as i32;

            // Add edges from colors inside the rectangle to color 'it'.
            for it2 in &st2 {
                graph[*it2 as usize].push(*it);
            }
        }

        // Step 4: Standard Topological Sort
        // Perform a topological sort using a Queue 'q' to check if there is a valid sequence of printing operations.
        let mut q: VecDeque<i32> = VecDeque::new();
        let mut count = 0;
        for it in &st {
            if indegree[*it as usize] == 0 {
                q.push_back(*it);
            }
        }
        while !q.is_empty() {
            let u = q.pop_front().unwrap();
            count += 1;
            for v in &graph[u as usize] {
                indegree[*v as usize] -= 1;
                if indegree[*v as usize] == 0 {
                    q.push_back(*v);
                }
            }
        }

        // Step 5: Checking the Result
        // If the 'count' variable is equal to the number of distinct colors in the 'targetGrid',
        // it means all colors were successfully processed in the topological sort, and we can return true.
        // Otherwise, we return false.
        count == st.len() as i32
    }
}
